prime-router/src/main/kotlin/azure/DownloadFunction.kt

```diff
@@ -0,0 +1,196 @@
+                    DAYS_TO_SHOW - it.task.createdAt.until(OffsetDateTime.now(), ChronoUnit.DAYS),
+                    it.task.itemCount,
+                    0,
+                    it.task.reportId.toString()
+                )
+            }
+    }
+
+    private fun responsePage(request: HttpRequestMessage<String?>, fileNotFound: Boolean = false): HttpResponseMessage {
+        val htmlTemplate: String = if( fileNotFound ) Files.readString( Path.of( FILENOTFOUND_PAGE ) ) else Files.readString( Path.of( DOWNLOAD_PAGE ) )
+        val headers = DatabaseAccess(dataSource = DatabaseAccess.dataSource).fetchHeaders(OffsetDateTime.now().minusDays(DAYS_TO_SHOW), orgName)
+
+        val attr = mapOf(
+            "description" to orgName,
+            "user" to userName,
+            "today" to Calendar.getInstance(),
+            "todays" to generateTodaysTestResults(headers),
+            "previous" to generatePreviousTestResults(headers),
+            "days_to_show" to DAYS_TO_SHOW,
+            "OKTA_redirect" to request.uri.toString()
+        )
+
+        val html = getTemplateFromAttributes(htmlTemplate, attr)
+
+        var response = request.createResponseBuilder(HttpStatus.OK)
+            .body(html)
+            .header("Content-Type", "text/html")
+            .build()
+
+        return response
+    }
+
+    private fun responseFile(request: HttpRequestMessage<String?>, fileName: String): HttpResponseMessage {
+        val header = DatabaseAccess(dataSource = DatabaseAccess.dataSource).fetchHeader(ReportId.fromString(fileName), orgName)
+        var response: HttpResponseMessage
+        try {
+            val body = WorkflowEngine().readBody(header)
+            if( body.size <= 0 )
+                response = responsePage( request, true );
+            else{
+            response = request
+                .createResponseBuilder(HttpStatus.OK)
+                .header("Content-Type", "text/csv")
+                .header("Content-Disposition", "attachment; filename=test-results.csv")
+                .body(body)
+                .build()
+            }
+        } catch (ex: Exception) {
+            response = request.createResponseBuilder(HttpStatus.NOT_FOUND).build()
+        }
+
+        return response
+    }
+
+    private fun getTemplateEngine(): TemplateEngine {
+        val templateEngine = TemplateEngine()
+        val stringTemplateResolver = StringTemplateResolver()
+        templateEngine.setTemplateResolver(stringTemplateResolver)
+        return templateEngine
+    }
+
+    private fun getTemplateFromAttributes(htmlContent: String, attr: Map<String, Any>): String {
+        val templateEngine = getTemplateEngine()
+        val context = Context()
+        if (!attr.isEmpty())
+            attr.forEach { (k, v) -> context.setVariable(k, v) }
+        return templateEngine.process(htmlContent, context)
+    }
+
+    private fun checkAuthenticated(request: HttpRequestMessage<String?>): Boolean {
+        userName = ""
```

Comment:
This doesn't look secure. Why can't an attacker forge a "cookie"?

Reply:
Updated to call JWTVerify - this takes the JWT token (stored as a cookie - becuase you cant set header properties on a redirect from the web page) and fetches the claims (sub & organizations) from Okta